### DISCUSSION TITLE
Livebook and its possible uses

### DISCUSSION THREAD

**Discussion by marcoherbst** [evercam/ex_nvr#534] [Ideas]
Wondering how best to take advantage of Livebook in combination with ExNVR. It seems to offer promise as a method of interactive documentation and possibly for internal tools (as per livebook teams). Potentially bundled or some other method for easy remote access and to execute code locally to the device.

Some possible uses:

- [ ] Create livebook templates (playbooks) to do common or device specific tasks that are not in the UI
- [ ] A playground for technical people to try some functionality or to familiarise themselves with ExNVR internals or to create small Proof of Concepts
- [ ] As above, the easy ability to share scripts for less common use cases e.g. local snapshot extraction. 
- [ ] Run https://github.com/acalejos/merquery as a way to explore with APIs on the camera. 
- [ ] Enable or disable WiFi on the device ( This is a an example playbook here: https://nerves-project.org/ ). This would allow the Pi to access internet without using a modem.
- [ ] (Experimental) features on a modem. Connect/Disconnect, Send SMS,
- [ ] Restart ExNVR https://github.com/evercam/ex_nvr/issues/357
- [ ] Reboot the Pi https://github.com/evercam/ex_nvr/issues/357
- [ ] Set Camera NTP settings https://github.com/evercam/ex_nvr/issues/34
- [ ] Update ExNVR to latest version
- [ ] Run an inference model on images 
- [ ] Tutorials for computer vision https://gist.github.com/ChristianAlexander/256f18818055494cbe4483390dd65314 
- [ ] Interactive ExNVR Documentation 
- [ ] Billal's "introduction to video" talk.
- [ ] https://github.com/membraneframework-labs/ex_vision 
- [ ] A Playbook to explore all OnVIF functionality
- [ ] Proof of Concept:  Read MPTT USB like we currently do on the router.
- [ ] Run a series of camera configuration API calls > A runbook for a particular camera config.
- [ ] Router Configuration > A runbook for the router or any other device on the network to configure it.
- [ ] Mount, UnMount, Format HDDs via a runbook.
- [ ] Recreate footage index.
- [ ] Some work was done recently with GPS on Teltonika. Could that be explained in a livebook? @AghilesBennia01 . Ask @JaviCT for an introduction to the basics and you will work out the rest.

Superficially at least, it seems the livebook teams people believe this sort of thing is reasonable:

![image](https://github.com/user-attachments/assets/843e70d3-e394-418c-9bef-ebfd24955382)

**Comment by the-prksh** [comment]
Hey 👋 ,

I am currently integrating `livebook ~> 0.13.3` and I have encountered a problem with dependency resolution. Livebook  depends on `phoenix_live_view ~> 1.0.0-rc.5`, while another project dependency, `flop_phoenix`, depends on `phoenix_live_view ~> 0.20.4`. `flop_phoenix` does not have upgrades available that support the required `phoenix_live_view`. 

I am struggling to find a way to resolve this conflict. Can anyone provide some insight on how to address this issue?

I added the following lines to `mix.exs` in the deps function:

```elixir
      {:livebook, "~> 0.13.2"},
      {:kino, "~> 0.13.2"},
      {:vega_lite, "~> 0.1"}
```

Here is the output of `mix` deps resolution:

```shell
. . .
Because your app depends on livebook ~> 0.13.2 which depends on phoenix_live_view 1.0.0-rc.5, phoenix_live_view 1.0.0-rc.5 is required.
So, because your app depends on phoenix_live_view ~> 0.20.0, version solving failed.
```

After upgrading the `phoenix_live_view` from `0.20.0` to `1.0.0-rc.5`
```shell
. . .
Because flop_phoenix >= 0.22.7 depends on phoenix_live_view ~> 0.20.4 and flop_phoenix >= 0.21.2 and < 0.22.7 depends on phoenix_live_view ~> 0.19.0 or ~> 0.20.0, flop_phoenix >= 0.21.2 requires phoenix_live_view ~> 0.19.0 or ~> 0.20.0.
And because your app depends on flop_phoenix ~> 0.22, phoenix_live_view ~> 0.19.0 or ~> 0.20.0 is required.
So, because your app depends on phoenix_live_view ~> 1.0.0-rc.5, version solving failed.
** (Mix) Hex dependency resolution failed
```

**Comment by gBillal** [comment]
I don't think including `livebook` as a dependency is a good idea. It's better to run a standalone livebook, creating the livebooks and connecting it to a running instance of `ex_nvr` to do the actual job.

If you still want to include it directly, either:  
* Upgrade the other dependencies to meet the requirements.
* Or find an older version of `livebook` that doesn't cause conflicts with the actual dependencies.

**Comment by the-prksh** [comment]
Running a standalone node locally and connecting it to a running instance of `ex_nvr` on a Raspberry Pi is not possible because `ex_nvr` runs on Elixir 1.15, while for Livebook the node must run Erlang/OTP 26 and Elixir 1.16.3 (or later). 

Either elixir or deps needs to be upgraded.

Here's a screenshot for reference: 

![Screen Shot 2024-08-14 at 3 49 41 PM](https://github.com/user-attachments/assets/613eaf10-cbc0-45cb-bebc-e22ac40b2da4)

**Comment by gBillal** [comment]
Unfortunately we cannot upgrade elixir to the latest version on `ex_nvr` for legacy reasons. However you can downgrade `livebook` to meet the requirements.

**Comment by the-prksh** [comment]
May I know the legacy reasons? 

I have upgraded the Elixir version in the Dockerfile to `hexpm/elixir:1.17.2-erlang-27.0-alpine-3.18.7` and everything seems to be working fine.

**Comment by the-prksh** [comment]
I have deployed and checked in Raspberry Pi as well, and it is working fine there as well. 

@gBillal Would you like to have a look?

**Comment by the-prksh** [comment]
When it comes to the Standalone livebook node in local machine and connecting runtime with edge device, 

It won't function on edge devices because the device doesn't have `mix` installed. Due to this, we are unable to install dependencies dynamically using `Mix.install/1`. 

You might be curious about how `nerves_livebook` addresses this issue. They tackle it by including all the `livebook` related dependencies as part of the release image and tweaking the `Mix.install/1` functionality. Here is the code for that - https://github.com/nerves-livebook/nerves_livebook/blob/main/lib/nerves_livebook/mix_install.ex



![Screen Shot 2024-08-16 at 3 57 45 PM](https://github.com/user-attachments/assets/3730317e-b3d7-417e-b54c-02d155e7d924)

**Comment by marcoherbst** [comment]
Do I understand correctly, that to use Livebook on a Pi, any dependencies would have to have been included in the firmware build? 

Is that true only if we were using nerves? (which at the moment we're not), or is this some other limitation related to a Pi that even with a regular OS (Debian, PiOS), Mix cannot be used?

**Comment by the-prksh** [comment]
It is true when integrating using an `Attached Node` or `Embedded Mode`, but only partially true when integrating using a `Standalone Node`.

It has nothing to do with nerves. The Pi OS will be trimmed compared to a regular OS. In our case, the Production release of `ex_nvr` will not contain Mix and other development dependencies.

This blog contains the difference between an `Attached node` and a `Standalone node`.  https://news.livebook.dev/remote-execution-smart-cell---launch-week-2---day-1-m3dv2

I said partially true because in `Standalone node`, The livebook node will run on the local machine and connect to the `ex_nvr` node. The notebook's dependencies will be installed on the local machine. It will use `erpc` to call functions of ex_nvr. However, it will not work with hardware functionalities such as Pi's peripherals, WiFi configurations, etc. To enable this functionality, an API needs to be developed in the firmware or the related dependencies should be added as part of the ex_nvr release.

**Comment by the-prksh** [comment]
I prefer using `embedded mode`. Nerves has used this approach to integrate Livebook. All dependencies will be included in the ex_nvr release, so there's no need to install anything on your local machine. Just use the link to open Livebook.

**Comment by gBillal** [comment]
Hi
The legacy reasons I'm talking about is that we're using devices (rasberry pi) with old debian versions. If we use Elixir 1.17.x to create the releases it'll include dependencies that'll not work on old versions (however an old version works fine on newer versions).

For livebook with attached mode, you cannot install dependencies, however you can do the following:

- Include the dependencies you need in exnvr itself and then use attached mode in livebook
- Or use smart cell (remote execution) to only get the data you need from exnvr and then use the livebook runtime to process that data. In this case you can use `System.cmd` or `System.shell` to execute commands on the remote machine (e.g. turn on/off wifi)

**Comment by the-prksh** [comment]
Hello @gBillal ,

Are you saying that some of the production kits contain old Debian versions that may cause issues? 

We are not changing the Debian version, only upgrading the Elixir/Erlang version. So, I don't think there would be any issue in that case. Is there any way we can test that?

Also, in `attached mode`, atleast elixir >= 1.16 is required.

**Comment by the-prksh** [comment]
> * Or use smart cell (remote execution) to only get the data you need from exnvr and then use the livebook runtime to process that data. In this case you can use `System.cmd` or `System.shell` to execute commands on the remote machine (e.g. turn on/off wifi)

This will be possible in "Standalone Mode", but for that, a livebook has to be set up on the laptop. @marcoherbst explained that he expects to integrate the livebook into the ex_nvr dashboard. This will not be possible if we go with `Standalone Mode`.

**Comment by gBillal** [comment]
> We are not changing the Debian version, only upgrading the Elixir/Erlang version. So, I don't think there would be any issue in that case. Is there any way we can test that?

Elixir 1.17 docker images are all based on the newer debian version and if that image is used to build arm64 releases, the external dependencies (like `ex_dtls`, `libsrtp2`, ..) will depend on newer version  of GLIBC which will not be loaded on old version of debian. I already tested this.

Anyway you can work on a new branch, upgrade elixir and any dependencies you need and create a release from there. Unfortunately we cannot upgrade right now.

**Comment by the-prksh** [comment]
@gBillal  Is it possible to upgrade Elixir to version `1.16` in that case? or Can you share the Debian version?

**Comment by gBillal** [comment]
As I said above, this work won't make it to the main branch. So you're free to work on a branch, upgrade to any version and create a release from there where we can test it on some devices.

we'll upgrade Elixir once we're no longer using old raspbian version on our side.

**Comment by marcoherbst** [comment]
Related: https://elixirforum.com/t/til-accessing-a-nerves-device-from-livebook/67028

**Comment by marcoherbst** [comment]
I wanted to note my understanding of ways this might work and my understandings of the pros & cons. My level of confidence on what I am saying here is very low. 

Also, I am writing purely from the perspecitve of an Evercam admin, which does happen to be ExNVR's ICP :) ... but I am not sure how generalisable this case is.

**Option 1 - Package Livebook with the ExNVR release**
This is not wholly different to how we have packaged https://github.com/phoenixframework/phoenix_live_dashboard and swagger API docs, though possibly on a different port. 
My understanding is that this is not really how livebook is intended to be used (option 2 is more traditional) but I see some evidence of people doing this for different reasons, e.g. https://github.com/nerves-livebook/nerves_livebook who use it as a demonstrator to get hands on with hardware.

**Option 2 - Run Livebook on my local PC and connect via VPN**
I'll call this the traditional method. Using this method I need to actively consider that I want my runbooks to execute on the remote device and I need a step to connect to the remote device. 

Level of user ability? 
Similar, though I think a simpler user might prefer option 1 as one doesn't need to consider where the code is being executed, it will always be done local to the device.

Security implications? 
A benefit of option 1 is that, in Evercam context, it does not require the user to become a client of the VPN.

Requires netbird / VPN to work? 
From an Evercam Admin PoV both options require netbird to be working, unless I happened to be on the same LAN as the NVR in which case Option 1 is preferred. 

Makes sense for an individual or fleet? 
I think option 1 makes sense in both cases, and I'm making the assumption that most use cases do not involve running a script accross many devices ... (in which case option 2 makes more sense).

Advantages over nervescloud scripts? 
More user friendly (subjective comment, but I'm the subject :)) as they can be better documented and the runbooks themselves form part of the documentation.

Implications re: Livebook Teams?
I dont know enough, but if I understand correctly, membership of a teams organisation could be a way of distributing access to certain livebooks. e.g. the evercam teams org would have premade, relevant evercam admin books.

Hoping this adds to the conversation. Happy to be told where I'm wrong.

**Comment by oussamabonnor1** [comment]
We had a successful experiment. Its not used anywhere but we can always go back to it.

Closing for now.
